src/libraries/Microsoft.Extensions.Configuration/src/ReferenceCountedProvidersManager.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.Configuration
+{
+    // ReferenceCountedProviderManager is used by ConfigurationManager to provide copy-on-write references that support concurrently
+    // reading config while modifying sources. It waits for readers to unreference the providers before disposing them
+    // without blocking on any concurrent operations.
+    internal sealed class ReferenceCountedProviderManager : IDisposable
+    {
+        private readonly object _replaceProvidersLock = new object();
+        private ReferenceCountedProviders _refCountedProviders = new(new List<IConfigurationProvider>());
+
+        // This is only used to support IConfigurationRoot.Providers because we cannot track the lifetime of that reference.
+        public IEnumerable<IConfigurationProvider> NonReferenceCountedProviders => _refCountedProviders.NonReferenceCountedProviders;
+
+        public ReferenceCountedProviders GetReference()
+        {
+            // Lock to ensure oldRefCountedProviders.Dispose() in ReplaceProviders() doesn't decrement ref count to zero
+            // before calling _refCountedProviders.AddRef().
+            lock (_replaceProvidersLock)
+            {
+                _refCountedProviders.AddReference();
+                return _refCountedProviders;
+            }
+        }
+
+        // Providers should never be concurrently modified. Reading during modification is allowed.
+        public void ReplaceProviders(List<IConfigurationProvider> providers)
+        {
+            ReferenceCountedProviders oldRefCountedProviders = _refCountedProviders;
+
+            lock (_replaceProvidersLock)
+            {
+                _refCountedProviders = new ReferenceCountedProviders(providers);
+            }
+
+            oldRefCountedProviders.Dispose();
+        }
+
+        public void AddProvider(IConfigurationProvider provider)
+        {
+            // Maintain existing references, but replace list with copy containing new item.
+            _refCountedProviders.Providers = new List<IConfigurationProvider>(_refCountedProviders.Providers)
+            {
+                provider
+            };
+        }
+
+        public void Dispose() => _refCountedProviders.Dispose();
```

Comment:
Should we assert that `_refCountedProviders` refcount is `0` after disposing it?

Reply:
I did this, but this made me consider how this assert (like all the others) isn't truly invariant if the ConfigurationManager is used after it's disposed. This indicates a bug in the code using the ConfigurationManager, but it would be fairly benign in most cases today. How do we feel about `Debug.Assert`s that are only valid of the API is used correctly? To me it feels like it's not invariant enough.

We could prevent this by preemptively throwing an ODE from `ReferenceCountedProviderManager.GetReference()` after it's disposed, but this might break existing apps that are today able to continue to read configuration after disposing a ConfigurationManager (or a ConfigurationRoot for that matter). I think disposing providers multiple times is the lesser evil.

We could also just start using a `ReferenceCountedProviders` instance that no longer reference counts after the ConfigurationManager is disposed. This way the `Debug.Assert`s are truly invariant and we only need to throw ODEs if something tries to change configuration sources after disposing the ConfigurationManager. I experimented with this [here](https://github.com/dotnet/runtime/compare/halter73/61747...halter73/61747-invariant?expand=1). I'm not sure if it's worth the complexity or not.

Reply:
> How do we feel about Debug.Asserts that are only valid of the API is used correctly? To me it feels like it's not invariant enough.

My thinking is that if a `Debug.Assert` can be made to fail using public APIs, then it really shouldn't be a `Debug.Assert`. @stephentoub might have a different opinion though...

For this case, if this was the first time we were shipping this type, I'd vote for throwing `ODE` if you used the object after you disposed of it. It seems like the "right" behavior to me. But I understand if we are concerned with breaking working applications.

If we choose to not throw ODE, then I think your latest commit makes the most sense from a behavior perspective at the cost of a bit of added complexity.

Reply:
> My thinking is that if a Debug.Assert can be made to fail using public APIs, then it really shouldn't be a Debug.Assert.

That is my stance as well.  Debug.Assert/Fail should only be used for things that should _never_ be able to happen; if you can trigger one, it's either a bug in the implementation or a faulty assert.